### PR TITLE
[GNA] Add support for SubstituteSoftSign for pattern with addition

### PR DESCRIPTION
**Pull request by aalbersk**
Jira: 41379

### CODE REVIEW

File: inference-engine/tests/functional/plugin/shared/src/subgraph_tests/softsign.cpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    Infer();
+
+    const auto& actualOutputs = GetOutputs();
+    auto referenceOutputs = CalculateRefs();
+
+    Compare(referenceOutputs, actualOutputs);
+}
+
+std::vector<std::vector<std::uint8_t>> SoftsignTest::CalculateRefs() {
+    InferenceEngine::Precision netPrecision = std::get<0>(this->GetParam());
```
Comment (bsochack) [inline inference-engine/tests/functional/plugin/shared/src/subgraph_tests/softsign.cpp:73]: Why is special reference function required?
If there is issue with precision, one of options is to modify threshold.

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp [hunk 2]
```diff
@@ -319,24 +319,53 @@ void SubstituteSoftSignPass::run() {
         }
         if (cont) continue;
         if (!hasNChildren(abs, 1)) continue;
-        auto power = getNthChild(abs, 0);
+        auto addition = getNthChild(abs, 0);
+        InferenceEngine::CNNLayerPtr power = nullptr;
+
+        if (!LayerInfo(addition).isPower()) continue;
+        auto powerLayer = LayerInfo(addition).as<PowerLayer*>();
+        if (powerLayer->power != -1) {
```
[the remark this answers is not in the document]
Comment (kmagiers) [reply inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:327]: also I think that more "compact" checks could be made to reduce number of "continue" paths
1) continue if scale || offset != 1 || abs(power) != 1
2) if(power == -1) do power assignment
3) else check child if is power
Much more compact and in my opinion flow is clearer

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp [hunk 3]
```diff
@@ -319,24 +319,53 @@ void SubstituteSoftSignPass::run() {
...
+                continue;
+            } else {
+                if (powerLayer->offset != 1) continue;
+                if (powerLayer->scale != 1) continue;
+
+                power = getNthChild(addition, 0);
+                if (!LayerInfo(power).isPower()) continue;
+                auto powerLayer_1 = LayerInfo(power).as<PowerLayer*>();
+                if (powerLayer_1->power != -1) continue;
+                if (powerLayer_1->offset != 0) continue;
```
Comment (kmagiers) [inline inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:340]: not explained, why child power offset must be 0 while original power offset must be 1?

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp [hunk 4]
```diff
@@ -319,24 +319,53 @@ void SubstituteSoftSignPass::run() {
...
+            if (powerLayer->offset != 1) continue;
+            if (powerLayer->scale != 1) continue;
 
-        if (!LayerInfo(power).isPower()) continue;
-        auto powerLayer = LayerInfo(power).as<PowerLayer*>();
-        if (powerLayer->power != -1) continue;
-        if (powerLayer->offset != 1) continue;
-        if (powerLayer->scale != 1) continue;
+            power = addition;
+            addition = nullptr;
```
Comment (kmagiers) [inline inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:348]: conside std::swap

File: inference-engine/tests/functional/plugin/gna/shared_tests_instances/subgraph_tests/softsign.cpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        {"GNA_COMPACT_MODE", "NO"}
+    }
+};
+
+std::vector<std::vector<size_t>> input_shapes = {
+    {1, 8},
+    {1, 42},
+    {1, 100},
+    {1, 128},
+    //{1, 1, 64},
```
Comment (kmagiers) [inline inference-engine/tests/functional/plugin/gna/shared_tests_instances/subgraph_tests/softsign.cpp:31]: to disable cases use skip_tests_config.cpp instead with proper JIRA ticket

**Review by kmagiers — CHANGES_REQUESTED** [review]

File: inference-engine/tests/functional/plugin/shared/src/subgraph_tests/softsign.cpp (hunk 1 above)
Comment (aalbersk) [reply inference-engine/tests/functional/plugin/shared/src/subgraph_tests/softsign.cpp:73]: There is no such operation as SoftSign in ngraph, so it is unable to check with that operation. There're also two Power operations: one in ngraph which is not used in inference engine conversion and one called PowerIE which is used there but is implemented in inference engine not ngraph, so ngraph interpreter doesn't recognize it. So I can do two things to implement the test: the one I did which is comparing the output only on GNA to see whether both patterns produce same results or implement the model with Power from ngraph and use this one in interpreter to compare model run on GNA with PowerIE and the one with Power run on ngraph interpreter.

Comment (kmagiers) [reply inference-engine/tests/functional/plugin/shared/src/subgraph_tests/softsign.cpp:73]: why not simply switch "function" variable after infer and before validate?

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (aalbersk) [reply inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:327]: fp32eq - done
merging ifs - done
cannot do abs(power) as if addition is added, power layer with power -1 still appears and need to differentiate

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp (hunk 3 above)
Comment (aalbersk) [reply inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:340]: Softsign is a computation that equals x/(1+|x|). Power operation used here is an operation which calculates: scale * x ^ power + offset [(here)](https://docs.openvinotoolkit.org/latest/openvino_docs_MO_DG_prepare_model_customize_model_optimizer_Subgraph_Replacement_Model_Optimizer.html). When addition is added separately, offset is set to 0.

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp [hunk 6]
```diff
@@ -319,24 +334,52 @@ void SubstituteSoftSignPass::run() {
...
+            if (!fp32eq(powerLayer->offset, 1) ||
+                !fp32eq(powerLayer->scale, 1)) continue;
 
-        if (!LayerInfo(power).isPower()) continue;
-        auto powerLayer = LayerInfo(power).as<PowerLayer*>();
-        if (powerLayer->power != -1) continue;
-        if (powerLayer->offset != 1) continue;
-        if (powerLayer->scale != 1) continue;
+            std::swap(addition, power);
+        }
```
Comment (kmagiers) [inline inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:363]: replace this nested if / else statemsents with:
```
// first layer after abs must have scale of 1, offset of 1 and power of either 1 or -1
if (!fp32eq(powerLayer.scale, 1.0f) || !fp32eq(powerLayer.offset, 1.0f) || !fp32eq(abs(powerLayer->power, 1.0f))) continue;
// power == -1, offset = 1, scale = 1
if(fp32eq(powerLayer->power, -1.0f)) {
    std::swap(addition, power);
} else { // power = 1, offset = 1, scale - 1 
    power = getNthChild(addition, 0);
    if (!LayerInfo(power).isPower()) continue;
    auto powerLayer_1 = LayerInfo(power).as<PowerLayer*>();
    // layer after addition must have power of -1, offset of 0 and scale of 1
    if (!(fp32eq(powerLayer_1->power, -1.0f) || !fp32eq(powerLayer_1->offset, 0.0f) || !fp32eq(powerLayer_1->scale, 1.0f)) continue;
}
```
The flow is much clearer this way and comments show exactly what to expect from each case

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp [hunk 7]
```diff
@@ -319,24 +334,52 @@ void SubstituteSoftSignPass::run() {
...
+                     << "\t" << power->name << "\n"
+                     << "\t" << mul->name << "\n"
+                     << std::endl;
+        else
+            gnalog() << "SoftSign subgraph found consits of: \n"
+                     << "\t" << abs->name << "\n"
+                     << "\t" << addition->name << "\n"
+                     << "\t" << power->name << "\n"
+                     << "\t" << mul->name << "\n"
+                     << std::endl;
```
Comment (kmagiers) [inline inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:382]: a lot of code repetition why not
```
gnalog() << "SoftSign subgraph found consits of: \n"
gnalog() << "\t" << abs->name << "\n"
if (addition != nullptr) gnalog() << "\t" << addition->name << "\n"
gnalog() << "\t" << power->name << "\n"
gnalog() << "\t" << mul->name << "\n"
gnalog() << std::endl;
```
Change is not obligatory though, seems like personal preference of mine

**Review by kmagiers — APPROVED** [review]

**Pull request merged**